### DISCUSSION TITLE
如何调试 Go 语言源代码 | Go 语言设计与实现 · /golang-debug

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#623] [Announcements]
https://draveness.me/golang/docs/part1-prerequisite/ch01-prepare/golang-debug/ 

 如何调试 Go 语言源代码

**Comment by shuff1e** [comment]
赞，可以出本书了

**Comment by ZhenhangTung** [comment]
分享在这个过程中我遇到的前置步骤：
1. [克隆源代码](https://golang.org/doc/install/source#fetch)并切到版本分支上
2. 运行`./src/make.bash`如果报错`make.bash must be run from $GOROOT/src`，可以在`/src`文件夹下运行`./make.bash`
3. 我添加`println("draven")`并没有显示`draven`字符串，一度以为我自己编译和使用出了问题，最后我添加了`Fprintln(os.Stdout, "draven")`能够显示确信我编译正确且使用了正确的编译包

**Comment by draveness** [comment]
> 分享在这个过程中我遇到的前置步骤：
> 
> 1. [克隆源代码](https://golang.org/doc/install/source#fetch)并切到版本分支上
> 2. 运行`./src/make.bash`如果报错`make.bash must be run from $GOROOT/src`，可以在`/src`文件夹下运行`./make.bash`
> 3. 我添加`println("draven")`并没有显示`draven`字符串，一度以为我自己编译和使用出了问题，最后我添加了`Fprintln(os.Stdout, "draven")`能够显示确信我编译正确且使用了正确的编译包

最后一个你确定么，我是可以打印的，你修改的是哪个函数？

**Comment by yanjinbin** [comment]
博主 什么时候出书 支持一把

**Comment by ZhenhangTung** [comment]
> > 分享在这个过程中我遇到的前置步骤：
> > 
> > 1. [克隆源代码](https://golang.org/doc/install/source#fetch)并切到版本分支上
> > 2. 运行`./src/make.bash`如果报错`make.bash must be run from $GOROOT/src`，可以在`/src`文件夹下运行`./make.bash`
> > 3. 我添加`println("draven")`并没有显示`draven`字符串，一度以为我自己编译和使用出了问题，最后我添加了`Fprintln(os.Stdout, "draven")`能够显示确信我编译正确且使用了正确的编译包
> 
> 最后一个你确定么，我是可以打印的，你修改的是哪个函数？

我尝试重复了整个操作，发现之前提供的信息有误。我的修改是否生效取决于我是否对`src/cmd/go/main.go`进行过修改。
1. 在我不改`src/cmd/go/main.go`的时候，我对`fmt/print.go`文件下的`func Sprintln(a ...interface{})`改动没有效果。
2. 修改了`src/cmd/go/main.go`之后，如我在`Sprintln`下加了个`println("cheers")`生效了。

Go版本位于`83b181c68b`，对应Go`1.14.4`。

## 没有改动`src/cmd/go/main.go`
编译：
```
➜  src git:(83b181c68b) ✗ ./make.bash
Building Go cmd/dist using /usr/local/Cellar/go/1.14.4/libexec. (go1.14.4 darwin/amd64)
Building Go toolchain1 using /usr/local/Cellar/go/1.14.4/libexec.
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
Building Go toolchain2 using go_bootstrap and Go toolchain1.
Building Go toolchain3 using go_bootstrap and Go toolchain2.
Building packages and commands for darwin/amd64.
---
Installed Go for darwin/amd64 in /Users/guiwu/Go-Code/src/goroot
Installed commands in /Users/guiwu/Go-Code/src/goroot/bin
```
运行：
```
➜  learn-go-src cat main.go
package main

import "fmt"

func main()  {
	fmt.Println("yeah!!!")
}
➜  learn-go-src /Users/guiwu/Go-Code/src/goroot/bin/go run main.go
yeah!!!
```

## 改动`src/cmd/go/main.go`后
`src/cmd/go/main.go`: 
```
func main() {
	_ = go11tag
	flag.Usage = base.Usage
	flag.Parse()
	log.SetFlags(0)

	fmt.Println("main function")
	# 剩余逻辑
}
```
`fmt/print.go`: 
```
// Println formats using the default formats for its operands and writes to standard output.
// S0paces are always added between operands and a newline is appended.
// It returns the number of bytes written and any write error encountered.
func Println(a ...interface{}) (n int, err error) {
	println("cheers")
	return Fprintln(os.Stdout, a...)
}
```
编译：
```
src git:(83b181c68b) ✗ ./make.bash
Building Go cmd/dist using /usr/local/Cellar/go/1.14.4/libexec. (go1.14.4 darwin/amd64)
Building Go toolchain1 using /usr/local/Cellar/go/1.14.4/libexec.
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
Building Go toolchain2 using go_bootstrap and Go toolchain1.
cheers
main function
Building Go toolchain3 using go_bootstrap and Go toolchain2.
cheers
main function
Building packages and commands for darwin/amd64.
cheers
main function
---
Installed Go for darwin/amd64 in /Users/guiwu/Go-Code/src/goroot
Installed commands in /Users/guiwu/Go-Code/src/goroot/bin
```
运行
```
➜  learn-go-src /Users/guiwu/Go-Code/src/goroot/bin/go run main.go
cheers
main function
yeah!!!
```

出现这个现象的原因我还未调研出来，不影响自己学习的方案就是在`src/cmd/go/main.go`里保留修改，比如打印一个空字符串，可以让自己学习下去，期间再慢慢深究原因。

**Comment by hapiman** [comment]
看了其他的几篇文章，非常值得学习。

**Comment by xiyichan** [comment]
如果要调试go语言源码中的编译器代码，即cmd/compile里面，不知道有何好办法，我用过gc和dlv但是很多过程都调试不出来，不知道如何使用。

**Comment by draveness** [comment]
> 如果要调试go语言源码中的编译器代码，即cmd/compile里面，不知道有何好办法，我用过gc和dlv但是很多过程都调试不出来，不知道如何使用。

你试过我这里的方法么，修改源代码然后构建出 Go 的二进制，用新的 Go 二进制编译源码

**Comment by xiyichan** [comment]
是使用./make.bash吧？这个我试过，但是有时候修改一些源码编译不通过，所以使用debug的方法看看就是什么函数负责什么功能。我现在往go提交PR也是用./make.bash编译好，然后用all.bash来测试。但是这样怎么说呢，我就想一步一步看代码怎么运行。



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Draven"<notifications@github.com&gt;; 
发送时间: 2020年9月16日(星期三) 上午8:43
收件人: "draveness/blog-comments"<blog-comments@noreply.github.com&gt;; 
抄送: "陈力锋"<2863768433@qq.com&gt;; "Comment"<comment@noreply.github.com&gt;; 
主题: Re: [draveness/blog-comments] 如何调试 Go 语言源代码 | Go 语言设计与实现 · /golang-debug (#215)





  
如果要调试go语言源码中的编译器代码，即cmd/compile里面，不知道有何好办法，我用过gc和dlv但是很多过程都调试不出来，不知道如何使用。
  
你试过我这里的方法么，修改源代码然后构建出 Go 的二进制，用新的 Go 二进制编译源码

**Comment by draveness** [comment]
> @ZhenhangTung 出现这个现象的原因我还未调研出来，不影响自己学习的方案就是在src/cmd/go/main.go里保留修改，比如打印一个空字符串，可以让自己学习下去，期间再慢慢深究原因。

说不定是因为编译缓存，不过我也不能确定

**Comment by xiyichan** [comment]
行吧，主要是想在cmd/compile这个包里面做点贡献，然后有很多文件的代码很难看懂有什么作用。我还没试过在cmd/go/里面做过修改。




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Draven"<notifications@github.com&gt;; 
发送时间: 2020年9月16日(星期三) 中午11:18
收件人: "draveness/blog-comments"<blog-comments@noreply.github.com&gt;; 
抄送: "陈力锋"<2863768433@qq.com&gt;; "Comment"<comment@noreply.github.com&gt;; 
主题: Re: [draveness/blog-comments] 如何调试 Go 语言源代码 | Go 语言设计与实现 · /golang-debug (#215)





  
@ZhenhangTung 出现这个现象的原因我还未调研出来，不影响自己学习的方案就是在src/cmd/go/main.go里保留修改，比如打印一个空字符串，可以让自己学习下去，期间再慢慢深究原因。
  
说不定是因为编译缓存，不过我也不能确定

**Comment by ClockworkTree** [comment]
请教下大佬，想运行标准库的test，但是有的包遇到了use of internal package net/http/internal not allowed的问题。如果包内没有引用internal的包就可以调试test，比如database，这个有什么好的解决方法吗

**Comment by draveness** [comment]
> 请教下大佬，想运行标准库的test，但是有的包遇到了use of internal package net/http/internal not allowed的问题。如果包内没有引用internal的包就可以调试test，比如database，这个有什么好的解决方法吗

你从外部引用 internal 包是不允许的

**Comment by ClockworkTree** [comment]
> > 请教下大佬，想运行标准库的test，但是有的包遇到了use of internal package net/http/internal not allowed的问题。如果包内没有引用internal的包就可以调试test，比如database，这个有什么好的解决方法吗
> 
> 你从外部引用 internal 包是不允许的

我是直接下载golang/go tag版本压缩文件，用goland直接打开下载的golang/go目录、来运行标准库里自己的test（不是我自己写的，从外部import的test），如果是非internal包，我直接对源码的改动，跑标准库自己的test是可以跑到的，说明test使用了goland打开的我自己下载源码而不是goroot里的sdk，但是遇到有internal的包，编译标准库test就会失败。

![image](https://user-images.githubusercontent.com/22406846/94809607-72210600-0425-11eb-8d5a-2dfb46811d1c.png)

**Comment by draveness** [comment]
> 我是直接下载golang/go tag版本压缩文件，用goland直接打开下载的golang/go目录、来运行标准库里自己的test（不是我自己写的，从外部import的test），如果是非internal包，我直接对源码的改动，跑标准库自己的test是可以跑到的，说明test使用了goland打开的我自己下载源码而不是goroot里的sdk，但是遇到有internal的包，编译标准库test就会失败。

我没有用过 Goland，大概看了下你提出的问题，我有两个疑问：

1. 使用命令行直接跑测试可以执行么
2. 编译错误的提示是什么，这里只看到了运行成功的输出

**Comment by benjamin-yanh** [comment]
汪明 是你？

**Comment by draveness** [comment]
> 汪明 是你？

不是

**Comment by tangximing** [comment]
> @ClockworkTree
> > > 请教下大佬，想运行标准库的test，但是有的包遇到了use of internal package net/http/internal not allowed的问题。如果包内没有引用internal的包就可以调试test，比如database，这个有什么好的解决方法吗
> > 
> > 你从外部引用 internal 包是不允许的
> 
> 我是直接下载golang/go tag版本压缩文件，用goland直接打开下载的golang/go目录、来运行标准库里自己的test（不是我自己写的，从外部import的test），如果是非internal包，我直接对源码的改动，跑标准库自己的test是可以跑到的，说明test使用了goland打开的我自己下载源码而不是goroot里的sdk，但是遇到有internal的包，编译标准库test就会失败。
> 
> ![image](https://user-images.githubusercontent.com/22406846/94809607-72210600-0425-11eb-8d5a-2dfb46811d1c.png)
> 

试试把GOPATH改一下，指向golang/go

**Comment by ClockworkTree** [comment]
@tangximing 
会报 internal not allowed的错误，需要把所有的internal的前面加个std/ 前缀，但是运行会遇到应该 golink 没有生效会有target找不到的错误

**Comment by ClockworkTree** [comment]
> @tangximing
> > @ClockworkTree
> > > > 请教下大佬，想运行标准库的test，但是有的包遇到了use of internal package net/http/internal not allowed的问题。如果包内没有引用internal的包就可以调试test，比如database，这个有什么好的解决方法吗
> > > 
> > > 你从外部引用 internal 包是不允许的
> > 
> > 我是直接下载golang/go tag版本压缩文件，用goland直接打开下载的golang/go目录、来运行标准库里自己的test（不是我自己写的，从外部import的test），如果是非internal包，我直接对源码的改动，跑标准库自己的test是可以跑到的，说明test使用了goland打开的我自己下载源码而不是goroot里的sdk，但是遇到有internal的包，编译标准库test就会失败。
> > 
> > ![image](https://user-images.githubusercontent.com/22406846/94809607-72210600-0425-11eb-8d5a-2dfb46811d1c.png)
> > 
> 
> 试试把GOPATH改一下，指向golang/go
> 

试了下，通过make.bash 重新编译后，将GOROOT指向重新编译后的目录是可以滴，GOPATH随意就好

**Comment by tangximing** [comment]
@ClockworkTree 我这边只设置了GOPATH就可以了
![image](https://user-images.githubusercontent.com/8757890/97653213-349eaf80-1a9b-11eb-8b72-3f882cbfcfff.png)

**Comment by gengjiawen** [comment]
出书+1，顺便问下 本网站 有GitHub链接吗？

**Comment by HeisenbergV** [comment]
用了好几天才勉强看完，写的太好了，谢谢，感觉水平提高了一个档次。

**Comment by gostudying** [comment]
图能看懂，涉及到源码的部分就看不懂了。我太菜了。

**Comment by polunzh** [comment]
代码块的高亮好像有问题，我在我浏览器上看不清:
![image.png](https://i.loli.net/2021/01/04/Kpb1f8OlVTEqJr7.png)

浏览器: Version 87.0.4280.88 (Official Build) (x86_64)

**Comment by draveness** [comment]
> 代码块的高亮好像有问题，我在我浏览器上看不清:
> ![image.png](https://camo.githubusercontent.com/0d56efdbe769db764470f831f6cf0ca5656f0d98a5a0705e75147c5a3cc01850/68747470733a2f2f692e6c6f6c692e6e65742f323032312f30312f30342f4b70623166384f6c565445714a72372e706e67)
> 
> 浏览器: Version 87.0.4280.88 (Official Build) (x86_64)

这个我也很奇怪，我本地也不行，但是很多人都可以，也不知道是啥问题，我在 Safari 没这个问题

**Comment by vclwei** [comment]
> @draveness
> > 代码块的高亮好像有问题，我在我浏览器上看不清:
> > ![image.png](https://camo.githubusercontent.com/0d56efdbe769db764470f831f6cf0ca5656f0d98a5a0705e75147c5a3cc01850/68747470733a2f2f692e6c6f6c692e6e65742f323032312f30312f30342f4b70623166384f6c565445714a72372e706e67)
> > 
> > 浏览器: Version 87.0.4280.88 (Official Build) (x86_64)
> 
> 这个我也很奇怪，我本地也不行，但是很多人都可以，也不知道是啥问题，我在 Safari 没这个问题
> 

我这里也是黑色的, 可能是 prism 的问题, 作者有时间看看?

**Comment by shizhz** [comment]
请教博主，有更方便的方式搭建本地环境来进行调试吗？如果只能每次都生成二进制文件的话，感觉整个流程好长啊

**Comment by Eason-Wan** [comment]
已经看到接口那里了，想问一下能下载到本地吗，方便线下学习

**Comment by sendme365** [comment]
平时用vscode debug 但是碰到一个情况就是当然input 输入是类似 （y/n） 时 。在debug steps时到这里后 就卡在这里了。因为找不到地方输入y 或者n 导致 debug steps 没有办法继续往下走。每次我只能hardcode 先。想问一下这样的情况怎么办？

**Comment by OnTheWay111** [comment]
MacOS上找不到dumped，也没找到安装方法，有遇到同样问题的小伙伴吗？

**Comment by draveness** [comment]
> 平时用vscode debug 但是碰到一个情况就是当然input 输入是类似 （y/n） 时 。在debug steps时到这里后 就卡在这里了。因为找不到地方输入y 或者n 导致 debug steps 没有办法继续往下走。每次我只能hardcode 先。想问一下这样的情况怎么办？

不知道，我没用 Debugger，Hardcode 也没啥问题靠猜

> MacOS上找不到dumped，也没找到安装方法，有遇到同样问题的小伙伴吗？

这里不是聊天室，建议给足上下文、清晰描述你的问题

**Comment by 0x1306a94** [comment]
可以在每一章的尾部添加上一页和下一页的功能吗?

**Comment by draveness** [comment]
> 可以在每一章的尾部添加上一页和下一页的功能吗?

已经实现了，可以看一下

**Comment by guoyuanchao1202** [comment]
段大师，我编译成功了，但是go run的时候报错了(windows平台下，安装的go版本是1.14，编译的go版本是最新的版本)，
代码是这样：
package main

import "fmt"

func main()  {
	fmt.Println("just test myGo Run")
}

错误大概是这样的：# sync/atomic
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:
  -D value
        predefined symbol with optional simple value -D=identifier=value; can be set multiple times
  -I value
        include directory; can be set multiple times
  -S    print assembly and machine code
  -V    print version and exit
  -debug
        dump instructions as they are parsed
  -dynlink
        support references to Go symbols defined in other shared libraries
  -e    no limit on number of errors reported
  -gensymabis
        write symbol ABI information to output file, don't assemble
  -newobj
        use new object file format
  -o string
        output file; default foo.o for /a/b/c/foo.s as first argument
  -shared
        generate code that can be linked into a shared library
  -trimpath string
        remove prefix from recorded source file paths
# internal/cpu
flag provided but not defined: -p
usage: asm [options] file.s ...
Flags:

**Comment by guoyuanchao1202** [comment]
感觉是不是我编译的版本(git上拉取的最新代码)和机器安装的(Go1.14.1)不兼容的问题？~

**Comment by lusir01** [comment]
突然发现没登录之前是早评论的在前面，登陆之后才是晩评论的在前面哈哈哈

**Comment by auula** [comment]
本篇已读打卡，继续下一篇，加油

**Comment by peihongch** [comment]
其实只需要进入src目录执行./make.bash就可以直接编译，使用的是你本机安装的go进行编译

**Comment by LiangquanLi930** [comment]
cool

**Comment by snail2sky** [comment]
看看  make.bash 脚本怎么写的 就知道 是啥问题了，而且这个网站有个问题，就是你写评论的时候，直接写 英文字母 s 会直接跳到搜索栏，咱换个组合键不好吗~~~

**Comment by snail2sky** [comment]
如果你机器上没有go环境，所以你得先装个go环境1，然后用这个Go去编译你刚修改的go源码

**Comment by Lincyaw** [comment]
我看不懂，但我大受震撼

**Comment by DamperHa** [comment]
博主，请问下，我按照你的方式修改了golang源码。但是有这么两个问题：
- 在fmt包的Println函数添加一条打印消息，照你的绝对路径运行并没有打印出我想要的消息
- 在runtime中makechan函数中，添加一条打印语句，却可以看到打印信息
这是为啥呀？

-=-

**Comment by iam153** [comment]
按照博主的方法操作了，啥也没看到。怀疑是go版本更新导致的？

**Comment by gongluck** [comment]
windows运行GOSSAFUNC=main go build main.go报错：'GOSSAFUNC' 不是内部或外部命令，也不是可运行的程序
或批处理文件。是命令不一样吗？

**Comment by sch-git** [comment]
> @ZhenhangTung
> > > 分享在这个过程中我遇到的前置步骤：
> > > 
> > > 1. [克隆源代码](https://golang.org/doc/install/source#fetch)并切到版本分支上
> > > 2. 运行`./src/make.bash`如果报错`make.bash must be run from $GOROOT/src`，可以在`/src`文件夹下运行`./make.bash`
> > > 3. 我添加`println("draven")`并没有显示`draven`字符串，一度以为我自己编译和使用出了问题，最后我添加了`Fprintln(os.Stdout, "draven")`能够显示确信我编译正确且使用了正确的编译包
> > 
> > 最后一个你确定么，我是可以打印的，你修改的是哪个函数？
> 
> 我尝试重复了整个操作，发现之前提供的信息有误。我的修改是否生效取决于我是否对`src/cmd/go/main.go`进行过修改。
> 1. 在我不改`src/cmd/go/main.go`的时候，我对`fmt/print.go`文件下的`func Sprintln(a ...interface{})`改动没有效果。
> 2. 修改了`src/cmd/go/main.go`之后，如我在`Sprintln`下加了个`println("cheers")`生效了。
> 
> Go版本位于`83b181c68b`，对应Go`1.14.4`。
> 
> ## 没有改动`src/cmd/go/main.go`
> 编译：
> ```
> ➜  src git:(83b181c68b) ✗ ./make.bash
> Building Go cmd/dist using /usr/local/Cellar/go/1.14.4/libexec. (go1.14.4 darwin/amd64)
> Building Go toolchain1 using /usr/local/Cellar/go/1.14.4/libexec.
> Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
> Building Go toolchain2 using go_bootstrap and Go toolchain1.
> Building Go toolchain3 using go_bootstrap and Go toolchain2.
> Building packages and commands for darwin/amd64.
> ---
> Installed Go for darwin/amd64 in /Users/guiwu/Go-Code/src/goroot
> Installed commands in /Users/guiwu/Go-Code/src/goroot/bin
> ```
> 运行：
> ```
> ➜  learn-go-src cat main.go
> package main
> 
> import "fmt"
> 
> func main()  {
> 	fmt.Println("yeah!!!")
> }
> ➜  learn-go-src /Users/guiwu/Go-Code/src/goroot/bin/go run main.go
> yeah!!!
> ```
> 
> ## 改动`src/cmd/go/main.go`后
> `src/cmd/go/main.go`: 
> ```
> func main() {
> 	_ = go11tag
> 	flag.Usage = base.Usage
> 	flag.Parse()
> 	log.SetFlags(0)
> 
> 	fmt.Println("main function")
> 	# 剩余逻辑
> }
> ```
> `fmt/print.go`: 
> ```
> // Println formats using the default formats for its operands and writes to standard output.
> // S0paces are always added between operands and a newline is appended.
> // It returns the number of bytes written and any write error encountered.
> func Println(a ...interface{}) (n int, err error) {
> 	println("cheers")
> 	return Fprintln(os.Stdout, a...)
> }
> ```
> 编译：
> ```
> src git:(83b181c68b) ✗ ./make.bash
> Building Go cmd/dist using /usr/local/Cellar/go/1.14.4/libexec. (go1.14.4 darwin/amd64)
> Building Go toolchain1 using /usr/local/Cellar/go/1.14.4/libexec.
> Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
> Building Go toolchain2 using go_bootstrap and Go toolchain1.
> cheers
> main function
> Building Go toolchain3 using go_bootstrap and Go toolchain2.
> cheers
> main function
> Building packages and commands for darwin/amd64.
> cheers
> main function
> ---
> Installed Go for darwin/amd64 in /Users/guiwu/Go-Code/src/goroot
> Installed commands in /Users/guiwu/Go-Code/src/goroot/bin
> ```
> 运行
> ```
> ➜  learn-go-src /Users/guiwu/Go-Code/src/goroot/bin/go run main.go
> cheers
> main function
> yeah!!!
> ```
> 
> 出现这个现象的原因我还未调研出来，不影响自己学习的方案就是在`src/cmd/go/main.go`里保留修改，比如打印一个空字符串，可以让自己学习下去，期间再慢慢深究原因。

遇上了和你相同的问题，解决方法也一样，go 的版本是1.16.8

**Comment by 018429** [comment]
谢谢博主，写的太好了~

**Comment by alpha-baby** [comment]
> @gongluck
> windows运行GOSSAFUNC=main go build main.go报错：'GOSSAFUNC' 不是内部或外部命令，也不是可运行的程序
> 或批处理文件。是命令不一样吗？

你应该去了解下再 类 unix 系统和 windows 系统中指定环境变量的区别

**Comment by SnowOnion** [comment]
> @OnTheWay111
> MacOS上找不到dumped，也没找到安装方法，有遇到同样问题的小伙伴吗？

`dumped` 那两行不是要执行的命令，是执行 `GOSSAFUNC=main go build main.go` 之后的输出 :D

**Comment by xiaobinqt** [comment]
```
Building Go cmd/dist using /usr/local/go. (go1.17.7 linux/amd64)
ERROR: $GOROOT_BOOTSTRAP must not be set to $GOROOT
Set $GOROOT_BOOTSTRAP to a working Go tree >= Go 1.4.
```

有个这个问题:cry:

**Reply by theten52** [reply]
看看我的评论

**Comment by theten52** [comment]
（Mac环境）按照楼主的文章一番操作很顺利的打印出了自定义字符。

无聊发现我自己的本地有两个一样版本的go可执行文件，一个是homebrew安装的，执行文件在/usr/local/bin/go下，实际是链接到/usr/local/Cellar/go/1.18.1/bin/go这个文件下的。另一个是使用官网安装包安装的。文件在/usr/local/go/bin/go。

刚开始，当我在homebrew安装的go的src目录下执行make.bash时一切正常，可以编译成功。然后我想把pkg安装go删掉，只保留homebrew安装的go（它们的版本都是一样的，1.18.1）。然而当我将官网安装包安装的go重命名后，再执行make.bash报错：ERROR: Cannot find /Users/xx/go1.4/bin/go.Set $GOROOT_BOOTSTRAP to a working Go tree >= Go 1.4.

调试了make.bash文件+查资料发现go1.16版本之后都需要另一个版本作为引导版本。恰好我的官网安装包安装的go充当了这个引导版本，才可以编译成功。make.bash的文件逻辑是找一个非当前go env中GOROOT变量下的go作用引导并设置为在GOROOT_BOOTSTRAP变量上。

如果有相同问题的同学可以尝试在系统中安装不同的go试试。

**Comment by zpeng2022** [comment]
博主你好，我第一次加print语句之后编译顺利通过了。但是第二次我在另外一处加了一行代码，之后运行make.bash，它一直卡在这里，
Building Go toolchain1 using /usr/local/Cellar/go/1.14.2_1/libexec.
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.

我是应该先uninstall 之前的go再重新编译吗？

**Comment by lazywhite** [comment]
本书是基于哪个版本的go写的呀? 建议标注下

**Reply by draveness** [reply]
1.15

**Reply by lazywhite** [reply]
建议在醒目的位置标注下

**Comment by hanniballei** [comment]
请问一下要学习1.18版本的go语言的话，这本书可以吗

**Comment by linyerun** [comment]
安装cloc工具：`sudo yum install cloc`
然后到go sdk的src所在目录运行：`cloc src`

**Comment by linyerun** [comment]
遇到的错误：
ERROR: Cannot find /root/go1.4/bin/go.
Set $GOROOT_BOOTSTRAP to a working Go tree >= Go 1.4.
然后我就没有编译，直接预先我的demo，发现我加的println("Ernie")生效了

**Comment by kiqi007** [comment]
关于1.2编译源码：

本人操作如下：
1. 使用go17编译go19原生代码(./all.bash)，获得二进制文件 ./bin/go，别名叫go19
(只编译这一次，未对源码做修改。   后面的测试均没再次编译)
2. 修改src/fmt/print.go下的方法，Println，其内容如文章增加一行打印
3. 使用步骤1中的go19去 go run 自写的./main.go文件，其中调用了Println方法。
4. 结果如文章，正确打印了：额外输出。

所以，go的源码编译和go的源码调试，实际上是两部分内容，并没有相互关系，不需要额外编译。
也就是说：go实际上，是使用$GoRoot/src的源码 + 用户源码，去构建新的二进制可执行文件而已。 而go19这个东西，只是编译/构建工具。

**Comment by xiaobinqt** [comment]
![](https://cdn.xiaobinqt.cn/xiaobinqt.io/20221220/efeed724728f412c88cd87109ec7a4df.png) 

编译成功了，但是执行有问题...

![](https://cdn.xiaobinqt.cn/xiaobinqt.io/20221220/6c0046d9ed8448368ef42e1c9027479d.png)

请问哪个大神有时间指导下，感谢。

**Comment by hopeJiangWang** [comment]
我是运行./src/make.bash报错make.bash must be run from $GOROOT/src。在/src文件夹下运行./make.bash，报错ERROR: Cannot find /home/hope/go1.4/bin/go.Set $GOROOT_BOOTSTRAP to a working Go tree >= Go 1.4.
最后是直接暴力：sudo rm -rf build_go &&  sudo  mkdir -p build_go 才可以

**Comment by LiaoSirui** [comment]
我是使用 g 工具管理多版本，运行时，还添加了

```bash
GOROOT=/code/go.googlesource.com/go /code/go.googlesource.com/go/bin/go run main.go
```

才能正确引用标准库

**Comment by YRiver6** [comment]
请教一下，windows Goland使用
set GOSSAFUNC=main
go build -gcflags="-S" main.go
仅仅生成了exe文件，ssa.html文件没有生成，怎么回事呀？

**Comment by Juminiy** [comment]
如何编译发布的第一个版本go1，我在公司服务器上试了试，工具链gcc11.4.0/clang14.0.0，sizeof()/ ^这个地方报错

**Reply by LiaoSirui** [reply]
截图下报错信息呗

**Comment by laurensent** [comment]
仅修改 `src/fmt/print.go` 编译后未生效的问题，排查后结论是：非编译缓存而是包指向性问题。如源代码仓库本地保存在 `/path/projects/go`，编译之后生成的二进制文件保存在 `/path/projects/go/bin`，使用如下命令
```bash
GOROOT=/path/projects/go/ /path/projects/go/bin/go  run main.go
```
或者在 IDEA，如 GoLand 中指定 Project GOPATH 后运行即可。

**Comment by yz-yz-yz** [comment]
添加println("draven")  会导致递归无线调用，  这里添加 
func Println(a ...any) (n int, err error) {
	Fprintln(os.Stdout, "DEBUG: Println called with", a)    
	return Fprintln(os.Stdout, a...)
}

**Comment by yz-yz-yz** [comment]
Go 分为两个部分：


标准库（std）	fmt、net/http、os 等库源码	你程序运行时依赖
工具链（cmd）	go、gofmt、compile 等工具的源码	编译器/工具行为改变才用

所以调试源码，不需要重新编译go 和 gofmt 工具本身。 直接进行修改源码运行即可。
